whatthelog/prefixtree/graph.py

```diff
@@ -45,6 +49,144 @@ def get_state_by_hash(self, state_hash: int):
             raise StateDoesNotExistException()
         return self.states[self.state_indices_by_hash[state_hash]]
 
+    def remove_loops(self, recurring: bool = False) -> None:
+        """
+        Main method to remove loops.
+        These can be recurring depending on the 'recurring' variable.
+        Please view the submethods for further explanation
+        :param recurring: determines whether recurring loops should be merged, or only singular loops
+                          (equivalent subsequent states).
+        """
+        if recurring:
+            self.__remove_recurrent_loops()
+        else:
+            self.__remove_singular_loops()
+
+    def __remove_singular_loops(self) -> None:
+        """
+        Removes equivalent subsequent loops from the graph.
+        For example if graph is:
+            0
+           /\
+          1  3
+          1  3
+          1  1
+          2
+          1
+        The output will be:
+            0
+           / \
+          ⊂1 ⊂3
+          |   |
+          2   1
+          |
+          1
+
+          ⊂ indicates a self-loop.
+        """
+        assert self.start_node is not None
+        current = self.start_node
+        been = set()
+        stack = [(current, been)]
+        while len(stack) > 0:
+            current, been = stack.pop()
+            outgoing = self.get_outgoing_states(current)
+            while len(outgoing) == 1:
+                if current.is_equivalent(outgoing[0]):
+                    self.merge_states(current, outgoing[0])
+                else:
+                    current = outgoing[0]
+                outgoing = [x for x in self.get_outgoing_states(current) if x is not current]
+            been.add(current)
+            for node in [n for n in outgoing if n not in been]:
+                stack.append((node, been.copy()))
+
+    def __remove_recurrent_loops(self) -> None:
+        """
+        Method that removes recurrent loops from branches of a graph.
+        For example if graph is:
+            0
+           /\
+          1  3
+          1  3
+          1  1
+          2
+          1
+        The output will be:
+            0
+           / \
+          ⊂1 ⊂3
+          ||  |
+          2   1
+
+          ⊂ indicates a self-loop and 2 has a edge from 1 to 2 and from 2 to 1.
+        """
+        assert self.start_node is not None
+        stack: List[Tuple[State, List]] = [(self.start_node, [])]
+        visited = set()
+
+        current_unique = []
+
+        while len(stack) != 0:
+            state, unique = stack.pop()
+            if len(unique) is not 0:
```

Comment:
```suggestion
            if len(unique) != 0:
```

The is operator should only be used for objects with the same reference.